webstruct/webannotator.py

```diff
@@ -280,11 +303,113 @@ def to_webannotator(tree, entity_colors=None, url=None):
     >>> wa_trees = [to_webannotator(tree, entity_colors) for tree in trees]  # doctest: +SKIP
 
     """
-    handler = _WaContentHandler(entity_colors)
-    lxml.sax.saxify(tree, handler)
-    tree = handler.out.etree
-    _copy_title(tree)
-    _add_wacolor_elements(tree, handler.entity_colors)
+    if not entity_colors:
+        entity_colors = EntityColors()
+
+    root = deepcopy(tree)
+
+    START_RE = re.compile(r' __START_(\w+)__ ')
+    END_RE = re.compile(r' __END_(\w+)__ ')
+
+    starts = list()
+    ends = list()
+    for _, element in etree.iterwalk(root, events=('start',)):
+
+        tasks = [(element.text, START_RE, starts, False),
+                 (element.text, END_RE, ends, False),
+                 (element.tail, START_RE, starts, True),
+                 (element.tail, END_RE, ends, True)]
+
+        for text, regexp, storage, is_tail in tasks:
+            if not text:
+                continue
+
+            for match in regexp.finditer(text):
+
+                if not match:
+                    continue
+
+                storage.append(TagPosition(element=element,
+                                           tag=match.group(1),
+                                           position=match.start(),
+                                           length=match.end() - match.start(),
+                                           is_tail=is_tail,
+                                           dfs_number=-1))
+
+    if len(ends) != len(starts):
+        raise ValueError('len(ends) != len(starts)')
+
+    # traverse tree in DFS manner
+    # each text is first child
+    # each tail is last
+    ordered = dict()
+    number = 0
+    for action, element in etree.iterwalk(root, events=('start', 'end')):
+
+        if action == 'end':
+            is_tail = True
+            ordered[(element, is_tail)] = number
+            number = number + 1  # for tail
+
+        if action == 'start':
+            is_tail = False
+            ordered[(element, is_tail)] = number
+            number = number + 1  # for text
+            number = number + 1  # for element
+
+    starts = [s for s in translate_to_dfs(starts, ordered)]
+    ends = [e for e in translate_to_dfs(ends, ordered)]
+
+    starts.sort(key=lambda t: (t.dfs_number, t.position))
+    ends.sort(key=lambda t: (t.dfs_number, t.position))
+
+    dfs_order = (max(n for n in ordered.values()) + 1) * [None]
+    for text_node, dfs_number in ordered.items():
+        dfs_order[dfs_number] = text_node
+
+    tasks = []
```

Comment:
Can we rename this variable, to make code more clear? Is it a list of positions to enclose with WebAnnotator spans? Something like positions_to_enclose would be more readable. There was even a variable with the same name within the same function, but with a different meaning.